pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageBrokerRequestHandler.java

```diff
@@ -247,6 +247,12 @@ private BrokerResponse handleRequest(long requestId, String query, @Nullable Sql
         // find a way to split metrics in case of multiple table
         String rawTableName = TableNameBuilder.extractRawTableName(tableNames.iterator().next());
         entry.getValue().setStageLevelStats(rawTableName, brokerResponseStats, _brokerMetrics);
+
+        // Track number of queries with number of groups limit reached
+        if (brokerResponse.isNumGroupsLimitReached()) {
```

Comment:
Say 3 tables were used in a join: `table_1, table_2, table_3`. If `table_2` was the one that had groups limit reached, the current logic will emit a metric for `table_2` and `table_3` since the same `brokerResponse` is written to in the loop.

I don't have context about this piece of code but I see `ExecutionStatsAggregation#setStats` is already emitting a bunch of metrics. I think we can emit this groups limit metric there as well.

The `rawTableName` is set to null right now which skips the metric emission there so we need to think about that.

Maybe we should emit global counters instead of table level counters there?

For Multistage queries, I am not sure if table level metrics make a lot of sense at broker level. My thought is that we can start with global metrics and then later think about table-level metrics.

@walterddr lmk your thoughts.

Reply:
@ankitsultana do you think it will make sense to emit a metric for just `table_2` here and skip for `table_3`. We can keep a track if the metric was emitted in one of the previous stages and skip in subsequent stages. I think this is better than emitting it for tableName as null. 
Let me know what do you think? cc @walterddr

Reply:
do we need to track exactly which table triggers the group limit? cuz the flag is global. there's really not much a user can do with per-table insight, correct?

Reply:
> do we need to track exactly which table triggers the group limit? cuz the flag is global. there's really not much a user can do with per-table insight, correct?

+1. I think we should simply go with global metrics. Table level metrics would be confusing if anything. We could consider adding table-level metrics later.

Reply:
> do we need to track exactly which table triggers the group limit? cuz the flag is global. there's really not much a user can do with per-table insight, correct?

Yeah this makes sense! We can just emit a global level metric in this case.

> I don't have context about this piece of code but I see ExecutionStatsAggregation#setStats is already emitting a bunch of metrics. I think we can emit this groups limit metric there as well.
The rawTableName is set to null right now which skips the metric emission there so we need to think about that.

@ankitsultana I think we cannot reuse this method as at every stage, this is called and we would end up emitting the metric twice again.

I think we can just maintain a flag that if in any one of the stages `isNumGroupsLimitReached` is true, we can emit a metric with `rawTableName` as null in this method itself outside the loop. 

@ankitsultana @walterddr does this sound good?

Reply:
Yeah you are right it would get called for every stage.

Can we do a general solution instead of emitting only `isNumGroupsLimitReached` metric? You can probably add a new method to ExecutionStatsAggregator and emit all the metrics that are being emitted for v1 right now, and we could call that at the end of the loop.

Reply:
oh. sorry i completely misunderstood your intent for this PR. so the goal is to emit a metrics not response metadata from broker returned to user?

Reply:
> oh. sorry i completely misunderstood your intent for this PR. so the goal is to emit a metrics not response metadata from broker returned to user?

Yes we want a metric to be emitted based on `isNumGroupsLimitReached`  field. We do it for v1 right now. 

> if that's the case you will have to attach some meaningful rawTableName here. otherwise simply a null is not differentiating for stats purpose.

As Ankit mentioned that we might emit metric twice if doing it inside the for loop. Should we save the first table we are getting this metric for and just emit based on that?

Reply:
just a quick thought. do we need to handle the "group-limit" metrics emission on broker? cant we do this on the servers and would that make more sense? 
- upon checking the group-by reducer code on broker it doesn't actually apply the group limit, 
- only place it is applied and trimmed result is on server group-by-operator and combine 
- v2 intermediate stage also will apply limit but at that point the "table" info is lost. 

does it make sense to directly emit the metrics on server?

Reply:
ok so the idea is to use default table or a opaque table if the groupLimitExceeded flag is set and unable to determine which table -- the benefit is to compute how many of these actually reached. but not to fine-grain control the signaling on which table.

I am ok with this as long as we are clear with the terminology